receiver/scraper.go

```diff
@@ -0,0 +1,120 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expvarreceiver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/expvarreceiver/internal/metadata"
+)
+
+type expVar struct {
+	// Use the existing runtime struct for decoding the JSON.
+	MemStats *runtime.MemStats `json:"memstats"`
+}
+
+type expVarScraper struct {
+	cfg        *Config
+	set        *component.ReceiverCreateSettings
+	httpClient *http.Client
+	mb         *metadata.MetricsBuilder
+}
+
+func newExpVarScraper(cfg *Config, set component.ReceiverCreateSettings) *expVarScraper {
+	return &expVarScraper{
+		cfg: cfg,
+		set: &set,
+		mb:  metadata.NewMetricsBuilder(cfg.MetricsConfig),
+	}
+}
+
+func (e *expVarScraper) start(_ context.Context, host component.Host) error {
+	httpClient, err := e.cfg.HTTPClientSettings.ToClient(host.GetExtensions(), e.set.TelemetrySettings)
+	if err != nil {
+		return err
+	}
+	e.httpClient = httpClient
+	return nil
+}
+
+func (e *expVarScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
+	emptyMetrics := pmetric.NewMetrics()
+	resp, err := e.httpClient.Get(e.cfg.Endpoint)
+	if err != nil {
+		return emptyMetrics, err
+	}
+	defer resp.Body.Close()
```

Comment:
There is a risk here that you may not actually have a response body and this could cause it to panic due to a nil reference, I would double check that you have a body before you continue here.

Reply:
This is how the official docs do it: https://pkg.go.dev/net/http#pkg-overview 

It also says: "If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close."